lib/std/crypto/aes.zig

```diff
@@ -0,0 +1,606 @@
+
+        var out: [exp_out.len]u8 = undefined;
+        var aes = AES256.init(key);
+        aes.decrypt(out[0..], in[0..]);
+        testing.expectEqualSlices(u8, exp_out[0..], out[0..]);
+    }
+}
+
+// Key expansion algorithm. See FIPS-197, Figure 11.
+fn expandKey(key: []const u8, enc: []u32, dec: []u32) void {
+    var i: usize = 0;
+    var nk = key.len / 4;
+    while (i < nk) : (i += 1) {
+        enc[i] = mem.readIntSliceBig(u32, key[4 * i .. 4 * i + 4]);
+    }
+    while (i < enc.len) : (i += 1) {
+        var t = enc[i - 1];
+        if (i % nk == 0) {
+            t = subw(rotw(t)) ^ (u32(powx[i / nk - 1]) << 24);
+        } else if (nk > 6 and i % nk == 4) {
+            t = subw(t);
+        }
+        enc[i] = enc[i - nk] ^ t;
+    }
+
+    var n = enc.len;
+    i = 0;
+    while (i < n) : (i += 4) {
+        var ei = n - i - 4;
+        var j: usize = 0;
+        while (j < 4) : (j += 1) {
+            var x = enc[ei + j];
+            if (i > 0 and i + 4 < n) {
+                x = td0[sbox0[x >> 24]] ^ td1[sbox0[x >> 16 & 0xff]] ^ td2[sbox0[x >> 8 & 0xff]] ^ td3[sbox0[x & 0xff]];
+            }
+            dec[i + j] = x;
+        }
+    }
+}
+
+test "expand key" {
+    const key = [_]u8{ 0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c };
+    const exp_enc = [_]u32{
+        0x2b7e1516, 0x28aed2a6, 0xabf71588, 0x09cf4f3c,
+        0xa0fafe17, 0x88542cb1, 0x23a33939, 0x2a6c7605,
+        0xf2c295f2, 0x7a96b943, 0x5935807a, 0x7359f67f,
+        0x3d80477d, 0x4716fe3e, 0x1e237e44, 0x6d7a883b,
+        0xef44a541, 0xa8525b7f, 0xb671253b, 0xdb0bad00,
+        0xd4d1c6f8, 0x7c839d87, 0xcaf2b8bc, 0x11f915bc,
+        0x6d88a37a, 0x110b3efd, 0xdbf98641, 0xca0093fd,
+        0x4e54f70e, 0x5f5fc9f3, 0x84a64fb2, 0x4ea6dc4f,
+        0xead27321, 0xb58dbad2, 0x312bf560, 0x7f8d292f,
+        0xac7766f3, 0x19fadc21, 0x28d12941, 0x575c006e,
+        0xd014f9a8, 0xc9ee2589, 0xe13f0cc8, 0xb6630ca6,
+    };
+    const exp_dec = [_]u32{
+        0xd014f9a8, 0xc9ee2589, 0xe13f0cc8, 0xb6630ca6,
+        0xc7b5a63,  0x1319eafe, 0xb0398890, 0x664cfbb4,
+        0xdf7d925a, 0x1f62b09d, 0xa320626e, 0xd6757324,
+        0x12c07647, 0xc01f22c7, 0xbc42d2f3, 0x7555114a,
+        0x6efcd876, 0xd2df5480, 0x7c5df034, 0xc917c3b9,
+        0x6ea30afc, 0xbc238cf6, 0xae82a4b4, 0xb54a338d,
+        0x90884413, 0xd280860a, 0x12a12842, 0x1bc89739,
+        0x7c1f13f7, 0x4208c219, 0xc021ae48, 0x969bf7b,
+        0xcc7505eb, 0x3e17d1ee, 0x82296c51, 0xc9481133,
+        0x2b3708a7, 0xf262d405, 0xbc3ebdbf, 0x4b617d62,
+        0x2b7e1516, 0x28aed2a6, 0xabf71588, 0x9cf4f3c,
+    };
+    var enc: [exp_enc.len]u32 = undefined;
+    var dec: [exp_dec.len]u32 = undefined;
+    expandKey(key[0..], enc[0..], dec[0..]);
+    testing.expectEqualSlices(u32, exp_enc[0..], enc[0..]);
+    testing.expectEqualSlices(u32, exp_dec[0..], dec[0..]);
+}
+
+// constants
+
+// TODO: generate these at comptime?
```

Comment:
Nah. If you were writing this code from scratch, that could make sense. But if you're porting an existing implementation, just match the port.